File: python/geojson-to-csv.py
import argparse
import json
import csv
import random
import string

def generate_random_hash(length):
    characters = string.ascii_letters + string.digits
    return ''.join(random.choice(characters) for _ in range(length))

def preprocess_column_name(column_name):
    # Replace hyphens and dollar signs with underscores
    return column_name.replace('-', '_').replace('$', '_')

def flatten_dict(data, prefix):
    flattened = {}
    for key, value in data.items():
        flattened[f"{prefix}__{key}"] = value
    return flattened

def main(input_filename, output_filename):
    with open(input_filename, 'r') as geojson_file:
        data = json.load(geojson_file)

    features = data['features']

    geometry_columns = set()
    properties_columns = set()

    for feature in features:
        geometry_columns.update(flatten_dict(feature['geometry'], 'g').keys())
        properties_columns.update(flatten_dict(feature['properties'], 'p').keys())

    with open(output_filename, 'w', newline='') as csv_file:
        csv_writer = csv.writer(csv_file)
        header = ['id'] + sorted(geometry_columns) + sorted(properties_columns)
        csv_writer.writerow([preprocess_column_name(col) for col in header])  # Preprocess header names

        for feature in features:
            if 'id' in feature:
                feature_id = feature['id']
            else:
                feature_id = generate_random_hash(16)
            
            geometry_fields = flatten_dict(feature['geometry'], 'g')
            properties_fields = flatten_dict(feature['properties'], 'p')
            
            row = [feature_id] + [geometry_fields.get(col, '') for col in sorted(geometry_columns)] + [properties_fields.get(col, '') for col in sorted(properties_columns)]
            csv_writer.writerow(row)

    print(f"CSV file '{output_filename}' has been successfully generated!")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Convert GeoJSON to CSV')
    parser.add_argument('--input', required=True, help='Input GeoJSON filename')
    parser.add_argument('--output', required=True, help='Output CSV filename')
    args = parser.parse_args()

    input_filename = args.input
    output_filename = args.output

    main(input_filename, output_filename)

File: node/geojson-to-csv.js
const fs = require('fs');
const path = require('path');
const argparse = require('argparse');
const csvWriter = require('csv-write-stream');

function generateRandomHash(length) {
  const characters = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
  let result = '';
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * characters.length));
  }
  return result;
}

function preprocessColumnName(columnName) {
  // Replace hyphens and dollar signs with underscores
  return columnName.replace(/-/g, '_').replace(/\$/g, '_');
}

function flattenObject(data, prefix) {
  let flattened = {};
  if (data !== null) {
    for (let [key, value] of Object.entries(data)) {
      flattened[`${prefix}__${key}`] = preprocessField(value);
    }
  }
  return flattened;
}

function preprocessField(value) {
  if (value === null || value === undefined) {
    return '';
  }
  return value.toString().replace(/\n/g, ' '); // Replace line breaks with spaces
}

function main(inputFilename, outputFilename) {
  const inputData = fs.readFileSync(inputFilename, 'utf8');
  const data = JSON.parse(inputData);

  const features = data.features;

  const geometryColumns = new Set();
  const propertiesColumns = new Set();

  for (let feature of features) {
    const geometryFields = flattenObject(feature.geometry, 'g');
    const propertiesFields = flattenObject(feature.properties, 'p');
    for (let key of Object.keys(geometryFields)) {
      geometryColumns.add(key);
    }
    for (let key of Object.keys(propertiesFields)) {
      propertiesColumns.add(key);
    }
  }

  const header = ['id', ...Array.from(geometryColumns).sort(), ...Array.from(propertiesColumns).sort()];
  const csvStream = csvWriter({
    headers: header.map(preprocessColumnName),
    separator: ',',
    newline: '\n',
  });

  const outputStream = fs.createWriteStream(outputFilename);
  csvStream.pipe(outputStream);

  for (let feature of features) {
    const featureId = feature.id || generateRandomHash(16);
    const geometryFields = flattenObject(feature.geometry, 'g');
    const propertiesFields = flattenObject(feature.properties, 'p');
    const row = [featureId, ...Array.from(geometryColumns).map((col) => geometryFields[col] || ''), ...Array.from(propertiesColumns).map((col) => propertiesFields[col] || '')];
    csvStream.write(row);
  }

  csvStream.end(() => {
    console.log(`CSV file '${outputFilename}' has been successfully generated!`);
  });
}

const parser = new argparse.ArgumentParser({
  description: 'Convert GeoJSON to CSV',
});
parser.addArgument(['--input'], { required: true, help: 'Input GeoJSON filename' });
parser.addArgument(['--output'], { required: true, help: 'Output CSV filename' });
const args = parser.parseArgs();

const inputFilename = args.input;
const outputFilename = args.output;

main(inputFilename, outputFilename);